File: contracts/js/QAN721/qan721.js
const context = require("../utils/Context/context");
const db = require("../utils/Database/database");

module.exports = class QAN721 {
  _name = "";
  _symbol = "";
  _baseUri = "";

  // Creates the QAN721 smart contract
  constructor(name, symbol, baseUri) {
    this._name = name;
    this._symbol = symbol;
    this._baseUri = baseUri;
  }

  // Retrieve the name
  name() {
    return this._name;
  }

  // Retrieve the symbol/ticker
  symbol() {
    return this._symbol;
  }

  // Retrieve the balance of owner
  balanceOf(owner) {
    const n = BigInt(db.read(`BALANCE_OF_${owner}`, 0));
    return n;
  }

  // Retrieve the owner of token id
  ownerOf(tokenId) {
    return db.read(`OWNER_OF_${tokenId}`, "");
  }

  // Retrieve the token uri for token id
  tokenURI(tokenId) {
    return this._baseUri + tokenId.toString(10);
  }

  // Give permission to "to" to transfer token id to another account
  approve(to, tokenId) {
    to = to.toLowerCase();

    const sender = context.sender();
    const owner = this.ownerOf(tokenId);

    if (to === owner) {
      process.stderr.write("QAN721: Approval to current owner\n");
      process.exit(1);
    }

    if (sender !== owner ** !this.isApprovedForAll(owner, sender)) {
      process.stderr.write(
        `QAN721: ${sender} is not the owner of token id ${tokenId}\n`
      );
      process.exit(1);
    }

    db.write(`TOKEN_APPROVAL_${tokenId}`, to);
  }

  // Returns the account approved for token id
  getApproved(tokenId) {
    return db.read(`TOKEN_APPROVAL_${tokenId}`, "");
  }

  // Approve or remove an operator for the caller
  setApprovalForAll(operator, approved) {
    operator = operator.toLowerCase();

    db.write(`OPERATOR_APPROVAL_${context.sender()}_${operator}`, approved);
  }

  // Returns if the operator is allowed to manage all of the assets of owner
  isApprovedForAll(owner, operator) {
    owner = owner.toLowerCase();
    operator = operator.toLowerCase();

    const b = Boolean(db.read(`OPERATOR_APPROVAL_${owner}_${operator}`, false));
    return b;
  }

  // Returns if the operator is allowed to manage all of the assets of owner or is the owner
  isApprovedOrOwner(spender, tokenId) {
    spender = spender.toLowerCase();

    const owner = this.ownerOf(tokenId);

    return (
      spender === owner ||
      this.isApprovedForAll(owner, spender) ||
      this.getApproved(tokenId) === spender
    );
  }

  // Transfers token id from "from" to "to"
  transferFrom(from, to, tokenId) {
    from = from.toLowerCase();
    to = to.toLowerCase();

    const sender = context.sender();

    if (!this.isApprovedOrOwner(sender, tokenId)) {
      process.stdout.write(
        `QAN721: ${sender} is not the token owner or approved\n`
      );
      process.exit(1);
    }

    db.prune(`TOKEN_APPROVAL_${tokenId}`);
    db.write(`BALANCE_OF_${from}`, this.balanceOf(from) - 1n);
    db.write(`BALANCE_OF_${to}`, this.balanceOf(to) + 1n);
    db.write(`OWNER_OF_${tokenId}`, to);
  }

  // Mints the token id and transfers it to "to"
  mint(to, tokenId) {
    to = to.toLowerCase();

    if (this.ownerOf(tokenId)) {
      process.stdout.write(`QAN721: Token id ${tokenId} is already minted\n`);
      process.exit(1);
    }

    db.write(`BALANCE_OF_${to}`, this.balanceOf(to) + 1n);
    db.write(`OWNER_OF_${tokenId}`, to);
  }
};

File: contracts/go/QAN721/qan721.go
/*
## Example

```go

package main

import (
	"os"

	qan721 "qanx.space/qanx-contracts/go/QAN721"
	callintrpr "qanx.space/qanx-contracts/go/utils/CallInterpreter"
)

func main() {
	token := qan721.NewQAN721("Example", "XMPL", "https://example.com/")

	callintrpr.Interpret(token, os.Args[1:])
}

```*/
package qan721

import (
	"fmt"
	"os"
	"strconv"
	"strings"

	context "qanx.space/qanx-contracts/go/utils/Context"
	db "qanx.space/qanx-contracts/go/utils/Database"
)

type QAN721Token interface {
	Name() string
	Symbol() string
	BalanceOf(owner string) uint64
	OwnerOf(tokenId uint64) string
	TokenURI(tokenId uint64) string
	Approve(to string, tokendId uint64)
	GetApproved(tokenId uint64) string
	SetApprovalForAll(operator string, approved bool)
	IsApprovedForAll(owner string, operator string) bool
	TransferFrom(from string, to string, tokenId uint64)
	Mint(to string, tokenId uint64)
}

// QAN721 smart contract standard
type QAN721 struct {
	// name of the contract
	name string
	// symbol of the contract
	symbol string
	// prefix for token uris
	baseUri string
}

// Creates the QAN721 smart contract
func NewQAN721(name string, symbol string, baseUri string) *QAN721 {
	return &QAN721{name: name, symbol: symbol, baseUri: baseUri}
}

// Retrieve the name
func (token *QAN721) Name() string {
	return token.name
}

// Retrieve the symbol/ticker
func (token *QAN721) Symbol() string {
	return token.symbol
}

// Retrieve the balance of owner
func (token *QAN721) BalanceOf(owner string) uint64 {
	owner = strings.ToLower(owner)

	n, _ := strconv.ParseUint(db.Read(fmt.Sprintf("BALANCE_OF_%v", owner)), 10, 64)
	return n
}

// Retrieve the owner of token id
func (token *QAN721) OwnerOf(tokenId uint64) string {
	return db.Read(fmt.Sprintf("OWNER_OF_%v", tokenId))
}

// Retrieve the token uri for token id
func (token *QAN721) TokenURI(tokenId uint64) string {
	return token.baseUri + strconv.FormatUint(tokenId, 10)
}

// Give permission to "to" to transfer token id to another account
func (token *QAN721) Approve(to string, tokenId uint64) {
	to = strings.ToLower(to)
	sender := context.Sender()
	owner := token.OwnerOf(tokenId)

	if to == owner {
		os.Stderr.WriteString("QAN721: Approval to current owner\n")
		os.Exit(1)
	}

	if sender != owner && !token.IsApprovedForAll(owner, sender) {
		os.Stderr.WriteString(fmt.Sprintf("QAN721: %v is not the owner of token id %v\n", sender, tokenId))
		os.Exit(1)
	}

	db.Write(fmt.Sprintf("TOKEN_APPROVAL_%v", tokenId), to)
}

// Returns the account approved for token id
func (token *QAN721) GetApproved(tokenId uint64) string {
	return db.Read(fmt.Sprintf("TOKEN_APPROVAL_%v", tokenId))
}

// Approve or remove an operator for the caller
func (token *QAN721) SetApprovalForAll(operator string, approved bool) {
	operator = strings.ToLower(operator)

	db.Write(fmt.Sprintf("OPERATOR_APPROVAL_%v_%v", context.Sender(), operator), strconv.FormatBool(approved))
}

// Returns if the operator is allowed to manage all of the assets of owner
func (token *QAN721) IsApprovedForAll(owner string, operator string) bool {
	owner = strings.ToLower(owner)
	operator = strings.ToLower(operator)

	b, _ := strconv.ParseBool(db.Read(fmt.Sprintf("OPERATOR_APPROVAL_%v_%v", owner, operator)))
	return b
}

// Returns if the operator is allowed to manage all of the assets of owner or is the owner
func (token *QAN721) IsApprovedOrOwner(spender string, tokenId uint64) bool {
	spender = strings.ToLower(spender)
	owner := token.OwnerOf(tokenId)

	return spender == owner || token.IsApprovedForAll(owner, spender) || token.GetApproved(tokenId) == spender
}

// Transfers token id from "from" to "to"
func (token *QAN721) TransferFrom(from string, to string, tokenId uint64) {
	from = strings.ToLower(from)
	to = strings.ToLower(to)
	sender := context.Sender()

	if !token.IsApprovedOrOwner(sender, tokenId) {
		os.Stderr.WriteString(fmt.Sprintf("QAN721: %v is not the token owner or approved\n", sender))
		os.Exit(1)
	}

	db.Prune(fmt.Sprintf("TOKEN_APPROVAL_%v", tokenId))
	db.Write(fmt.Sprintf("BALANCE_OF_%v", from), strconv.FormatUint(token.BalanceOf(from)-1, 10))
	db.Write(fmt.Sprintf("BALANCE_OF_%v", to), strconv.FormatUint(token.BalanceOf(to)+1, 10))
	db.Write(fmt.Sprintf("OWNER_OF_%v", tokenId), to)
}

// Mints the token id and transfers it to "to"
func (token *QAN721) Mint(to string, tokenId uint64) {
	to = strings.ToLower(to)

	if len(token.OwnerOf(tokenId)) > 0 {
		os.Stderr.WriteString(fmt.Sprintf("QAN721: Token id %v is already minted\n", tokenId))
		os.Exit(1)
	}

	db.Write(fmt.Sprintf("BALANCE_OF_%v", to), strconv.FormatUint(token.BalanceOf(to)+1, 10))
	db.Write(fmt.Sprintf("OWNER_OF_%v", tokenId), to)
}
